### PR TITLE
WSearchLineEdit: Add options to disable auto-completion and history

### LINKED ISSUE #10634: Add option to disable search completions
Reported by: **[fwcd](https://launchpad.net/~fwcd)**
Date: 2022-01-04T21:42:30Z
Status: Confirmed
Importance: Low
Launchpad Issue: [lp1956392](https://bugs.launchpad.net/bugs/1956392)
Tags: gui, library

---

It would be nice to have a setting for disabling the completions in the library search bar.

Version: 2.4.0-alpha-pre
Revision: bf1eef2e26

### PR DESCRIPTION
**Pull request by fwcd**
### Fixes #10634

This adds two options (that are enabled by default) to configure whether search completions and history (including up/down keyboard shortcuts) should be enabled:

image

### CODE REVIEW

**Comment by fwcd** [comment]
Hm, the build failure is strange, is `setAutoCompletion` not available on all platforms?

**Comment by fwcd** [comment]
It's mostly minor things that I find a bit inconvenient:

- The completions kicking in too quickly, as you've described in the issue
- Accidentally hitting up/down or turning the track selection knob on the controller now moves through the history. While this of course is the intended behavior, I find it a bit irritating since I now have to seek back through the history and figure out what I previously wanted to type
- I don't really use the feature, so being able to disable it makes the UI a bit less noisy

**Comment by ronso0** [comment]
Just a side note, the new options should be in the Library page.

**Comment by ronso0** [comment]
> Figure out why setAutoCompletion doesn't compile on all platforms

Use [`setCompleter(completer*)`](https://doc.qt.io/qt-5/qcombobox.html#setCompleter)?
Either `completer*` is a valid QCompleter or a nullptr (disables auto-completion as described in the doc).

Also, instead of parsing the config on each Up/Down/Space press, why not use a static functions to set bools, like the one setting the debounce timeout?
[void WSearchLineEdit::setDebouncingTimeoutMillis(int debouncingTimeoutMillis)](https://github.com/mixxxdj/mixxx/blob/49164318d384c1c92d3bf73cac527c1af50af289/src/widget/wsearchlineedit.cpp#L63-L66)

Furthermore, I'm not comfortable with removing the history function altogether (which is not the case now btw, queries are still saved and restored, though not accessible).
Isn't it sufficient for your use case to suppress the emulated keystrokes [sent by librarycontrol](https://github.com/mixxxdj/mixxx/blob/49164318d384c1c92d3bf73cac527c1af50af289/src/library/librarycontrol.cpp#L664-L680), i.e. still allow regular hotkeys (Up, Down, Ctrl+Space) and keep the expand button?

**Comment by fwcd** [comment]
I've moved the settings as per your suggestion to the Library page (see the updated screenshot in the PR description) and renamed the history setting to 'Enable search history shortcuts' to clarify that it only affects the shortcuts and not the history itself.

Suppressing the emulated keystrokes in `LibraryControl` would probably work but IMO feels a bit like the wrong place to model this (if we consider the search history to be an implementation detail of the search bar, we should also handle it there and not swallow the key events at a higher level). On the other hand, the generated keys are already somewhat tied to the search bar's semantics. Not sure about this one.

Regarding the static functions: While this may work for the history shortcuts (since we only need to read the value during key events), `setSearchCompletionsEnabled` cannot really be static if we want to to swap the completer based on whether search completions are enabled. The todo comment mentions that this also is more of a workaround currently and that it would be better to wire up the config changes with signals and slots. Since we would have access to the `WSearchLineEdit` instance that way, this would also solve the aforementioned issue.

Since I'm not that familiar with the data flow of configuration changes yet, do you have some pointers on how to do cleanly propagate config changes to `WSearchLineEdit`?

**Comment by ronso0** [comment]
> `setSearchCompletionsEnabled` cannot really be static if we want to to swap the completer based on whether search completions are enabled.

You could check if this bool changed in the focusInEvent()

**Comment by ronso0** [comment]
Anyways, I assume this setting is changed rarely, so applying this after start / skin change/reload may be sufficient IMO

**Comment by fwcd** [comment]
I've now wired up the settings as static bools as you suggested. Also, `WSearchLineEdit` now creates a custom owned `QCompleter` that is set depending on whether search completions are enabled.

One thing that struck me as slightly odd is that some settings are updated differently: For example, the debounce time is updated immediately when the user changes the widget whereas most other settings are first applied when the user clicks 'Apply'. This feels a bit inconsistent (but is probably a topic for another PR). With these new settings, I've tried to match the original semantics (i.e. first update the config when 'Apply' is clicked).

In the longer term, we'd probably want to find a clean way to notify `WSearchLineEdit` about config changes, the updating-static-variables feels like a workaround to a deeper architectural issue currently.

**Comment by ywwg** [comment]
I support this feature, I do not personally like the autocomplete / history but I understand others may find it useful.

File: src/widget/wsearchlineedit.cpp
Comment (ronso0) [inline src/widget/wsearchlineedit.cpp:703]: undefined, and not useful here

File: src/widget/wsearchlineedit.h
Comment (ronso0) [inline src/widget/wsearchlineedit.h:3]: why not `<QCompleter>` like the other Qt includes?

File: src/widget/wsearchlineedit.h
Comment (ronso0) [inline src/widget/wsearchlineedit.h:84]:
```suggestion
    static bool s_searchCompletionsEnabled;
    static bool s_searchHistoryShortcutsEnabled;
```
just because without 'search' prefix readability in the cpp is better IMO, and less redundant e.g. when getting it from within the preferences `WSearchLineEdit::kEnableCompletionsByDefault`

File: src/preferences/dialog/dlgpreflibrarydlg.ui
Comment (ronso0) [inline src/preferences/dialog/dlgpreflibrarydlg.ui:387]: the checkboxes should be added to `<tabstops>` near the end of the file

**Review by ronso0 — CHANGES_REQUESTED** [review]
I'm sorry you had to wait so long, this was kinda pushed off the table.

It's working nicely, thanks. Just some nitpicks.
I guess it doesn't matter that the completer is always created even if it will never actually be loaded to lineEdit->completer.

File: src/widget/wsearchlineedit.h
Comment (ronso0) [inline src/widget/wsearchlineedit.h:28]:
```suggestion
    static constexpr bool kEnableSearchCompletionsByDefault = true;
    static constexpr bool kEnableSearchHistoryShortcutsByDefault = true;
```
because 'search' is redundant e.g. when getting it from within the preferences `WSearchLineEdit::kEnableCompletionsByDefault`

**Review by ronso0 — CHANGES_REQUESTED** [review]

**Comment by ronso0** [comment]
I think it works also with the "Search history:" label (also we need less space and thereby potentially avoid horizontal scrollbars for long translations)
image

- [ ] Enable search history **keyboard** shortcuts ?

File: src/widget/wsearchlineedit.h
Comment (fwcd) [reply src/widget/wsearchlineedit.h:3]: This was what clangd auto-included it under, but I agree that `<QCompleter>` is cleaner and more consistent

**Comment by fwcd** [comment]
I've updated the branch with your suggestions

File: src/widget/wsearchlineedit.cpp
Comment (ronso0) [inline src/widget/wsearchlineedit.cpp:5]: Oh, during my previous review I didn't notice this is already included in wsearchlineedit.h
Please remove it here, git commit --amend (maybe name it more explicitely) and push one last time.

FYI rebasing after a review _without approval_ is not appreciated since that means someone has to review again, instead of just checking the last commits. Here, it's not much of a drama because the change set is rather small and can be verified by manual testing.

Comment (fwcd) [reply src/widget/wsearchlineedit.cpp:5]: > FYI rebasing after a review without approval is not appreciated

Ah sorry, I'll try to keep that in mind for the next time

Comment (ronso0) [reply src/widget/wsearchlineedit.cpp:5]: I didn't neither review again nor test manually, so if you like to --fixup previous commits with the latest changes and squash..  feel free.

Comment (fwcd) [reply src/widget/wsearchlineedit.cpp]: I went with amending for now, most of these are just small stylistic changes anyway

**Comment by fwcd** [comment]
Any updates on this?

**Comment by daschuer** [comment]
After https://github.com/mixxxdj/mixxx/pull/11207 is merged, the completion is no longer applied immediately. It requires "→" or "⏎" to take effect. 

This means the 

> The completions kicking in too quickly, as you've described in the issue

part will be solved. 

Can we reduce this to only add the "EnableSearchHistoryShortcuts" preferences option for less clutter in the preferences?

File: src/widget/wsearchlineedit.h
Comment (daschuer) [inline src/widget/wsearchlineedit.h:27]: This is a function name, how about 
```suggestion
    static constexpr bool kDefaultSearchCompletionsEnabled = true;
    static constexpr bool kDefaultSearchHistoryShortcutsEnabled = true;
```

**Comment by fwcd** [comment]
I still think having the option to disable auto-completion would be cool, if only to reduce visual noise

File: src/preferences/dialog/dlgpreflibrary.cpp
Comment (daschuer) [inline src/preferences/dialog/dlgpreflibrary.cpp:516]: Since this is not a slot, name the function according to what it does, not when it is called. 
```suggestion
void DlgPrefLibrary::updateSearchLineEditHistoryOptions() {
```

**Review by ronso0 — APPROVED** [review]

**Review by daschuer — CHANGES_REQUESTED** [review]

**Comment by daschuer** [comment]
> I still think having the option to disable auto-completion would be cool, if only to reduce visual noise

OK

File: src/widget/wsearchlineedit.h
Comment (ronso0) [reply src/widget/wsearchlineedit.h:27]: [I preferred to remove](https://github.com/mixxxdj/mixxx/pull/10942#pullrequestreview-1189077067) the redundant 'search' string, so let's name these
`kCompletionsEnabledDefault`
`kHistoryShortcutsEnabledDefault`
?

**Review by daschuer — APPROVED** [review]
LGTM Thank you.

**Pull request merged**
